src/ir_passes/new_mono.rs

```diff
@@ -0,0 +1,868 @@
+use crate::ir::{self, Ctx, Foreign, IndexStore, MutCtx};
+use itertools::Itertools;
+use linked_hash_map::LinkedHashMap;
+use std::collections::HashMap;
+
+/// Represents a deferred monomorphize instance that will be processed by the pass at a later time.
+struct MonoDeferred<'a, 'pass: 'a> {
+    /// The name of the monomorphized component
+    base: ir::Component,
+    /// The underlying component to be monomorphized
+    underlying: &'a ir::Component,
+    /// Mapping from sig-owned parameters in the underlying component to their constant bindings.
+    binding: ir::Bind<ir::ParamIdx, u64>,
+    /// Mapping from non-sig-owned params in the underlying component to the parameters in the new component
+    /// that they've been replaced with
+    par_binding: ir::Bind<ir::ParamIdx, ir::ParamIdx>,
+
+    /// Underlying pointer
+    pass: &'a mut Monomorphize<'pass>,
+
+    /// Mapping of underlying invokes (and how many times we've seen it) to base invokes
+    inv_map: HashMap<(ir::InvIdx, u32), ir::InvIdx>,
+    /// Mapping from underlying invokes to how many times we've seen it so far
+    inv_counter: HashMap<ir::InvIdx, u32>,
+    /// Mapping of underlying instances (and how many times we've seen it) to base instances
+    inst_map: HashMap<(ir::InstIdx, u32), ir::InstIdx>,
+    /// Maapping from underlying instances to how many times we've seen it so far
+    inst_counter: HashMap<ir::InstIdx, u32>,
```

Comment:
I'm not sure I've followed what these counters are used for. I see that we count up but why do we need to do that?